sycl/doc/design/VirtualFunctions.md

```diff
@@ -0,0 +1,428 @@
+    // can't decide whether both 'foo' and 'bar' should be both included in the
+    // resulting device image - the decision must be made at runtime when we
+    // know the target device.
+    new (Obj) Base;
+  });
+
+  // The same binary produced by thy sycl compiler should correctly work on both
+  // devices with and without support for 'fp64' aspect.
+  Q.single_task<Use>(syclext::properties{syclext::calls_indirectly<>}, [=] {
+    Obj->foo();
+  });
+
+  if (Q.get_device().has(sycl::aspect::fp64)) {
+    Q.single_task<Use>(syclext::properties{syclext::calls_indirectly<set_fp64>},
+        [=] {
+      Obj->bar();
+    });
+  }
+
+  return 0;
+}
+```
+
+As comments in the snippet say the main issue is with vtables: at compile time
+it may not be clear which exact functions can be safely included in there and
+which are not in order to avoid speculative compilation and fulfill optional
+kernel features requirements from the SYCL 2020 specification.
+
+To solve this, the following approach is used: all virtual functions marked with
+`indirectly_callable` property are grouped by set they belong to and outlined
+into separate device images (i.e. device images with kernels using them are left
+with declarations only of those virtual functions).
+
+For each device image with virtual functions that use optional features we also
+create a "dummy" version of it where bodies of all virtual functions are
+emptied.
+
+Dependencies between device images are recorded in properties based on
+`calls_indirectly` and `indirectly_callable` properties. They are used later by
+runtime to link them together. Device images which depend on optional kernel
+features are linked only if those features are supported by a target device and
+dummy versions of those device images are used otherwise.
+
+This way we can emit single unified version of LLVM IR where vtables reference
+all device virtual functions, but their definitions are outlined and linked
+back dynamically based on device capabilities.
+
+For AOT flow, we don't do outlining and dynamic linking, but instead do direct
+cleanup of virtual functions which are incompatible with a target device.
+
+## Design
+
+### Changes to the SYCL header files
+
+New compile-time properties `indirectly_callable` and `calls_indirectly` should
+be implemented in accordance with the corresponding [design document][2]:
+
+- `indirectly_callable` property should lead to emission of
+  `"indirectly-callable"="set"` function attribute, where "set" is a string
+  representation of the property template parameter.
+- `calls_indirectly` property should lead to emission of
+  `"calls-indirectly"="set1,set2"`, where "set1" and "set2" are string
+  representations of the property template parameters.
+
+In order to convert a type to a string, [\__builtin_sycl_unique_stable_name][3]
+could be used.
+
+**TODO**: `calls_indirectly` requires compile-time concatenation of strings.
+Document how it should be done.
```

Comment:
@intel/dpcpp-cfe-reviewers, I would like to hear your feedback here.

Even though it is possible to implement compile-time string concatenation using meta-programming magic and maybe even possible to make it work on non-literals using `constexpr` `strlen` built-in clang provides, I don't think that it should be a way to go. It would be way too complicated, fragile and will impact compilation time which is slow enough already with SYCL.

Instead, I was thinking about extending `__builtin_sycl_unique_stable_name` to accept more than one `typename`, so it returns a string which contains comma-separated list of those unique stable names. Do you think that this is something we can implement? How hacky does it look? How acceptable that is considering upstreaming efforts? Would it be possible to keep it as a customization without upstreaming?

Reply:
We're trying to move away from `__builtin_sycl_unique_stable_name` in the case where Clang is the host compiler, but we'll need to retain it for third party host compiler support. That suggests we'll end up with, effectively, two implementations of this new functionality. The one needed for upstreaming is still being investigated, but is expected to use IR introspection to accomplish the same goals.

For the non-upstream implementation, rather than extend `__builtin_sycl_unique_stable_name`, I think it would be preferable to introduce a new builtin; `__builtin_sycl_unique_stable_name_list` or similar.

Reply:
I have figured out a way how to achieve the desired effect of converting `calls_indirecltly<int, void>` into `"calls-indirectly"="_ZSTi,_ZSTv"` without need of any extra compiler built-ins. The solution is documented in f33bab59e165de42149cad747abb28bb447bbbe1.

`__builtin_sycl_unique_stable_name` will continue to be used to convert typenames into string literals, but I imagine if there would be a new solution for that task, then we will simply switch to it here as well. The usage model of `calls_indirectly` property parameters is very similar to kernel names: it is used to help runtime identify some bits in device images, which are referenced by user through typenames.